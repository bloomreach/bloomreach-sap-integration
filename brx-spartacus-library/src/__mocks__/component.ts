import { Component as SpaSdkComponent } from '@bloomreach/spa-sdk';

export class MockComponent implements SpaSdkComponent {
  getId = jest.fn();

  getMeta = jest.fn();

  getModels = jest.fn();

  getUrl = jest.fn();

  getName = jest.fn();

  getParameters = jest.fn();

  getProperties = jest.fn();

  getChildren = jest.fn();

  getComponent = jest.fn();

  getComponentById = jest.fn();
}

export default new MockComponent();
