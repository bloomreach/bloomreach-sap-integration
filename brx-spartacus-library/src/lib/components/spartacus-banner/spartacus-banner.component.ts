/*
 * Copyright 2020-2021 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component, ComponentFactoryResolver, Injector, Input, OnInit, ViewChild } from '@angular/core';
import { ContainerItem, Document, getContainerItemContent, ImageSet, Page } from '@bloomreach/spa-sdk';
import { CmsBannerComponent } from '@spartacus/core';
import { BannerComponent, CmsComponentData } from '@spartacus/storefront';
import { of } from 'rxjs';
import { BannerCompound } from '../../models/common-types.model';
import { SpartacusBannerDirective } from './spartacus-banner.directive';

@Component({
  selector: 'brx-spartacus-banner',
  templateUrl: './spartacus-banner.component.html',
})
export class SpartacusBannerComponent implements OnInit {
  @Input() component!: ContainerItem;

  @Input() page!: Page;

  @ViewChild(SpartacusBannerDirective, { static: true })
  wrappedComponent!: SpartacusBannerDirective;

  constructor(private readonly componentFactoryResolver: ComponentFactoryResolver) {}

  ngOnInit(): void {
    this.renderWrappedBanner();
  }

  async renderWrappedBanner(): Promise<void> {
    const componentFactory = this.componentFactoryResolver.resolveComponentFactory(BannerComponent);
    const { viewContainerRef } = this.wrappedComponent;
    viewContainerRef.clear();

    const banner = getContainerItemContent<BannerCompound>(this.component, this.page);
    const image = banner?.image && this.page.getContent<ImageSet>(banner.image);
    const link = banner?.link && this.page.getContent<Document>(banner.link);
    const content = await this.page.rewriteLinks(banner?.content?.value ?? '');

    const componentProperties: CmsBannerComponent = {
      headline: banner?.title,
      content,
      media: {
        url: image?.getOriginal()?.getUrl(),
      },
      urlLink: link?.getUrl(),
      styleClasses: banner?.styleClasses,
    };

    const componentInjector = Injector.create({
      providers: [{ provide: CmsComponentData, useValue: { data$: of(componentProperties) } }],
    });

    viewContainerRef.createComponent<BannerComponent>(componentFactory, 0, componentInjector);
  }
}
