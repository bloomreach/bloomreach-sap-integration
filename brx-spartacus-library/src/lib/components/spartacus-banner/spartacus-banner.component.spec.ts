import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterModule } from '@angular/router';
import { APP_BASE_HREF } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, ViewContainerRef } from '@angular/core';
import { Page } from '@bloomreach/spa-sdk';
import sanitizeHTML from 'sanitize-html';
import mockPage from '../../../__mocks__/page';
import mockComponent from '../../../__mocks__/container-item';
import { SpartacusBannerComponent } from './spartacus-banner.component';
import { SpartacusBannerDirective } from './spartacus-banner.directive';

jest.mock('sanitize-html', jest.fn);

describe('SpartacusBannerComponent', () => {
  let component: SpartacusBannerComponent;
  let componentEl: HTMLElement;
  let fixture: ComponentFixture<SpartacusBannerComponent>;

  const mockWrappedComponent = {
    viewContainerRef: {
      clear: jest.fn(),
      createComponent: jest.fn(),
    },
  };

  const mockImageUrl = 'imageURL';
  const mockLinkUrl = 'linkURL';

  const mockContentDataImage = {
    getOriginal: jest.fn(() => ({ getUrl: jest.fn(() => mockImageUrl) })),
  };

  const mockContentDataLink = {
    getUrl: jest.fn(() => mockLinkUrl),
  };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [RouterModule],
      declarations: [SpartacusBannerComponent],
      providers: [{ provide: APP_BASE_HREF, useValue: '/' }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      teardown: { destroyAfterEach: false },
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SpartacusBannerComponent);
    component = fixture.componentInstance;
    componentEl = fixture.nativeElement;

    component.component = {
      ...mockComponent,
    };

    component.page = {
      ...mockPage,
      isPreview: jest.fn(() => false),
      getContent: jest.fn((content) => {
        if (content === 'image') {
          return mockContentDataImage;
        }
        if (content === 'link') {
          return mockContentDataLink;
        }
        return {
          type: 'componentcontent',
          data: {
            image: 'image',
            link: 'link',
            content: { value: 'value' },
            title: 'title',
            styleClasses: 'styleClasses',
          },
        };
      }),
    } as unknown as Page;

    component.wrappedComponent = new SpartacusBannerDirective(
      mockWrappedComponent.viewContainerRef as unknown as ViewContainerRef,
    );

    // eslint-disable-next-line no-import-assign
    (sanitizeHTML as any) = jest.fn();
    (sanitizeHTML.defaults as any) = {
      allowedTags: [],
    };

    fixture.detectChanges();
  });

  it('should initialize component correctly', () => {
    jest.spyOn(component, 'renderWrappedBanner');

    component.ngOnInit();

    fixture.detectChanges();

    expect(component.renderWrappedBanner).toHaveBeenCalled();
    expect(mockWrappedComponent.viewContainerRef.createComponent).toHaveBeenCalled();
  });

  it('should initialize when page content is empty', () => {
    jest.spyOn(component.page, 'getContent').mockReturnValueOnce({});

    component.renderWrappedBanner();

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
