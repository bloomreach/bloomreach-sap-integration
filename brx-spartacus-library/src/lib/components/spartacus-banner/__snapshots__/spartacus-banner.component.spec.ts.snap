// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpartacusBannerComponent should initialize when page content is empty 1`] = `
<div
  id="root1"
>
  <div>
    
  </div>
</div>
`;
